File: app/models/course_number.rb
class CourseNumber
  class MalformedCourseNumberError < StandardError; end

  REGEX = %r{PBS\s?CH\s?(?<cantonal_association>[a-zA-Z]{2})\s?(?<kind>\d)(?<regional_association>\d)(?<count_number>\d)-(?<year>\d{2})}

  attr_reader :cantonal_association, :regional_association, :year, :kind, :count_number

  def initialize(course_number_string)
    match = REGEX.match(course_number_string.to_s) || {}
    @course_number_string = course_number_string
    @cantonal_association = match[:cantonal_association]
    @regional_association = match[:regional_association]
    @count_number = match[:count_number]
    @year = 2000 + match[:year].to_i
    @kind = match[:kind]

    raise MalformedCourseNumberError.new unless valid?
  end

  def valid?
    cantonal_association && regional_association && count_number && year && kind
  end

  def pbs_ch?
    @course_number_string&.start_with?('PBS CH')
  end

  def to_s
    @course_number_string
    # "PBS CH #{cantonal_association} #{kind}#{regional_association}#{count_number}-#{year}"
  end
end

File: src/models/course_number.ts
export class NoCourseNumberError extends Error {
}

export type CourseNumber = {
  cantonalAssociation?: string;
  regionalAssociation: string;
  year: number
  kind: number
  countNumber: number
}

export function parseCourseNumber(courseNumberString: string): CourseNumber {
  const regex = new RegExp(/PBS\s?(CH)?\s([a-zA-Z]{2})?\s?(?<kind>\d)(?<regionalAssociation>\d)(?<countNumber>\d)-(?<year>\d{2})/, 'i')
  const match = regex.exec(courseNumberString)
  if (!match) throw new NoCourseNumberError('Not a course number')

  return {
    cantonalAssociation: match[2]?.toUpperCase(),
    kind: +match[3],
    regionalAssociation: match[4],
    countNumber: +match[5],
    year: +match[6],
  }
}

export function getAssociation(courseNumber: CourseNumber): string {
  if (!courseNumber.cantonalAssociation || courseNumber.cantonalAssociation === '') return 'CH';
  if (courseNumber.cantonalAssociation === 'ZH') return `ZH-${courseNumber.regionalAssociation}`;
  return courseNumber.cantonalAssociation || '';
}

export function formatCourseNumber(courseNumber: CourseNumber): string {
  const cantonalAssociation = courseNumber.cantonalAssociation ? `${courseNumber.cantonalAssociation} ` : ''
  return `PBS CH ${cantonalAssociation}${courseNumber.kind}${courseNumber.regionalAssociation}${courseNumber.countNumber}-${courseNumber.year}`
}
